### PR TITLE
Synethetic Source: Fix scaled float

### PR DESCRIPTION
**Pull request by nik9000**
This causes scaled float values that entirely saturate their numeric
range to continue saturating their range on round trip.

### CODE REVIEW

File: modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/ScaledFloatFieldMapper.java [hunk 1]
```diff
@@ -677,7 +677,54 @@ public SourceLoader.SyntheticFieldLoader syntheticFieldLoader() {
...
+                if (scaledValue == Long.MIN_VALUE) {
+                    double min = Long.MIN_VALUE / scalingFactor;
+                    if (Math.round(min * scalingFactor) != Long.MIN_VALUE) {
+                        double v = min - Math.ulp(min);
+                        assert Math.round(v * scalingFactor) == Long.MIN_VALUE;
+                        return v;
+                    }
+                    return min;
+                }
+                return scaledValue / scalingFactor;
```
Comment (romseygeek) [inline modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/ScaledFloatFieldMapper.java:727]: Given this logic is fairly gnarly, is it worth extracting it into a helper and then throwing some heavy-duty testing on it? Even something quick-check like, to test boundary conditions and then a solid random selection of points.

Comment (nik9000) [reply modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/ScaledFloatFieldMapper.java:727]: We get that from the round trip tests but I could certainly see wanting it to be tested more directly. I'll yank it around a bit.

**Comment by nik9000** [comment]
@romseygeek I've pushed some more tests. Let me know if you think I should add even more. This is a tricky thing.

**Review by romseygeek — APPROVED** [review]
LGTM. Let's get it in and see what CI makes of it.

File: modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/ScaledFloatFieldMapper.java [hunk 2]
```diff
@@ -677,8 +681,55 @@ public SourceLoader.SyntheticFieldLoader syntheticFieldLoader() {
...
+     * {@link Long#MIN_VALUE} for {@code double}s that are too small. But
+     * {@code Long.MAX_VALUE / scalingFactor} doesn't always yield a value
+     * that would saturate. In other words, sometimes:
+     * <pre>{@code
+     *   long scaled1 = Math.round(BIG * scalingFactor);
+     *   assert scaled1 == Long.MAX_VALUE;
+     *   double decoded = scaled1 / scalingFactor;
+     *   long scaled2 = Math.round(decoded * scalingFactor);
+     *   assert scaled2 != Long.MAX_VALUE;
+     * }</pre>
```
Comment (not-napoleon) [inline modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/ScaledFloatFieldMapper.java:708]: Good explanation, thanks for including it.

**Review by not-napoleon — APPROVED** [review]
LGTM.  Or, well, looks as good as this much floating point jank can look ;)  Nice work.

**Comment by nik9000** [comment]
Be free PR! Go and mingle with your brothers and sisters in CI and find bugs.

**Pull request merged**